object.c

```diff
@@ -4677,6 +4699,14 @@ InitVM_Object(void)
     rb_define_method(rb_cFalseClass, "===", case_equal, 1);
     rb_undef_alloc_func(rb_cFalseClass);
     rb_undef_method(CLASS_OF(rb_cFalseClass), "new");
+
+    rb_mImmutable = rb_define_module("Immutable");
+    rb_define_method(rb_mImmutable, "new", rb_immutable_new, -1);
+    rb_define_singleton_method(rb_mImmutable, "extended", rb_immutable_extended, 1);
+
+    rb_mImmutableObject = rb_define_module_under(rb_mImmutable, "Object");
+    rb_define_method(rb_mImmutableObject, "dup", rb_immutable_obj_clone, -1);
+    rb_define_method(rb_mImmutableObject, "clone", rb_immutable_obj_clone, -1);
```

Comment:
This is not enough as mentioned in https://bugs.ruby-lang.org/issues/18035?next_issue_id=18034&prev_issue_id=18036#note-21.
I think better to not need `Immutable::Object` at all, but rather have Kernel#{clone/dup} have a way to know all instances of a class are immutable.

Reply:
Can't anyone just override `dup` and `clone`?

Reply:
It's not enough as it's still possible to call Kernel's clone/dup and actually get mutable instances (see issue).

Reply:
Do you think we need to worry about people who intentionally blow their foot off?

Reply:
In this case I think yes. It might also enable using Immutable for core types instead of the hardcoded list which seems nice.